nodejs.py

```diff
@@ -0,0 +1,234 @@
+"""
+Utility-belt script to manage the Node.js dependency
+"""
+import argparse
+import collections
+import fnmatch
+import glob
+import io
+import os
+import sys
+import tarfile
+import zipfile
+
+import build_system.constants as c
+import build_system.build_constants as bc
+import build_system.build_utils as utils
+import build_system.build_settings as settings
+
+# helper classes to show zipping progress
+# original idea: https://stackoverflow.com/a/3668977/425532
+class ReadProgressFileObject(io.FileIO):
+    current_read = 0
+    def __init__(self, path, *args, **kwargs):
+        io.FileIO.__init__(self, path, *args, **kwargs)
+
+    def read(self, size):
+        b = io.FileIO.read(self, size)
+        ReadProgressFileObject.current_read += len(b)
+        return b
+
+class WriteProgressFileObject(io.FileIO):
+    def __init__(self, path, size, *args, **kwargs):
+        self._total_size = size
+        io.FileIO.__init__(self, path, *args, **kwargs)
+
+    def write(self, b):
+        progress = min(100.0, ReadProgressFileObject.current_read / (self._total_size * 0.01))
+        sys.stdout.write("\r[%3.2f%%] " %(progress))
+        sys.stdout.flush()
+        return io.FileIO.write(self, b)
+
+Command = collections.namedtuple("Command", "aliases function")
+DepsDirectory = collections.namedtuple("DepsDirectory", "path include")
+
+# Command-Line setup
+parser = argparse.ArgumentParser(formatter_class=argparse.RawTextHelpFormatter)
+
+#-----------------------------------------------------------------------
+def flush_cache(silent = False):
+    if not silent:
+        print "[flush-cache]"
+
+    utils.store_nodejs_output(None, ".")
+
+    if not silent:
+        print "Done" 
+
+cmd_flush_cache = Command(
+    aliases=["flush-cache", "fc"],
+    function=flush_cache,
+)
+#-----------------------------------------------------------------------
+def git_init():
+    print "[git-init]"
+
+    # TODO: add CLI overide options
+    # - Node version
+    # - J2V8 version
+
+    utils.store_nodejs_output(None, ".")
+
+    if (not os.path.exists("node")):
+        print "Cloning Node.js version: " + settings.NODE_VERSION
+        # NOTE: autocrlf=false is very important for linux based cross-compiles of Node.js to work on a windows docker host
+        utils.execute("git clone https://github.com/nodejs/node --config core.autocrlf=false --depth 1 --branch v" + settings.NODE_VERSION)
```

Comment:
Currently the build.py script will not download the Node.js code if it is missing. That is what the `nodejs.py` script is intended for, and also because of `separation of concerns` I deliberately decided against mixing the management tasks for build dependencies (Node.js/V8) with the actual build process. That is also the reason why there are seperate scripts for those two tasks. In the upcoming commits I will add some more documentation to BUILDING.md how to run a full build off everything starting from a blank checkout.